[Torch Package] Make get names of OrderedImporters support fallback to importers

dl541 opened the pull request:
Summary:
OrderedImporters is supposed to be an importer which tries out every single importer in self._importers. However the get_name API does not follow this behavior and only uses the get_name from the basic Importer class.
This change is to update the OrderedImporters get_name API so that it tries the get_name API of every single importers.

Differential Revision: D76463252

jingsh reviewed (approved):
make sense in general, left some questions

dl541 commented on torch/package/importer.py:
```diff
@@ -204,6 +205,14 @@ def _is_torchpackage_dummy(self, module):
         return module.__file__ is None
 
+    def get_name(self, obj: Any, name: Optional[str] = None) -> tuple[str, str]:
+        for importer in self._importers:
+            try:
+                return importer.get_name(obj, name)
+            except (ObjNotFoundError, ObjMismatchError) as e:
+                logging.warning(f"Tried to call get_name with obj {obj} and name {name} on {importer} and got {e}")
```
The behavior of OrderedImporters is to try all the importers one-by-one until one succeeds. The existing function import_module already has this behavior. If a module exists in more than one of the importers, we will just use the first importer that succeeds.

My implementation of get_name exhibits the same property. We will use the first importer that is able to do get_name successfully.

jingsh commented:
@pytorchbot merge

dl541 committed:
[Torch Package] Make get names of OrderedImporters support fallback to importers (#155743)

Summary:

OrderedImporters is supposed to be an importer which tries out every single importer in self._importers. However the get_name API does not follow this behavior and only uses the get_name from the basic Importer class.
This change is to update the OrderedImporters get_name API so that it tries the get_name API of every single importers.

Reviewed By: seanx92, jingsh

Differential Revision: D76463252

dl541 commented:
The linux-jammy-py3_9-clang9-xla test is a false alarm. The jax dependency issue is being addressed in #158876 pull request.

izaitsevfb commented:
@pytorchbot merge -i
